Support Promises for imageClassifier

tmcw opened the pull request:
This includes:

- The return values of imageClassifier() and ImageClassifier.predict()
  are now promises.
- Uncomments, updates, and implements tests for ImageClassifier

### The changes

ml5 classes have traditional constructors but many of them have asynchronous 'loading' requirements and then synchronous methods. ES6 class constructors can't be async methods (you can't say `await new Foo()`), so this does the following:

- All classes that need some async setup phase have a `ready` property that is a Promise that is resolved when they're ready
- The 'factory functions' that people use to get instances of those classes (like the imageClassifier class for ImageClassifier) have two behaviors:
  - The user provides a callback and they return the class instance
  - The user does not provide a callback and they return the class instance's 'ready' property, which is a promise that resolves to that class instance once it's ready.

So:

```js
// callbacks
var ifier = imageClassifier('MobileNet', undefined, {}, onReady);
function onReady() {
  // do stuff with ifier
}

// promises
imageClassifier('MobileNet', undefined, {}).then(iffier => {
  // do stuff with iffier
});

// async/await syntax
async function run() {
  let iffier = await imageClassifier('MobileNet', undefined, {});
  // do stuff with iffier
}
```

**Other changes**

- Implements tests for everything that's testable!
  - This doesn't include ImageClassifier. The `@tf-models/mobilenet` module that powers ImageClassifier doesn't run in the karma/webpack environment for reasons I can't quite ascertain right now.
- Switches the karma reporter from progress, which is very minimal, to mocha, which is a bit more human. Karma re-runs portions of your tests and, especially developing new tests, I wanted to know which tests were running and which had passed/failed, instead of only getting details on failures and only getting summary numbers.

<details>
<summary>:book: The error around mobilenet</summary>

Error message:

```js
    Error: Argument 'x' passed to 'conv2d' must be a Tensor, but got object.
        at c (src/ImageClassifier/index.test.js:1:2672)
        at a (src/ImageClassifier/index.test.js:1:2126)
        at n (src/ImageClassifier/index.test.js:1:2330)
        at Object.i (src/ImageClassifier/index.test.js:1:2355)
        at e.conv2d (src/ImageClassifier/index.test.js:1:25150)
        at src/ImageClassifier/index.test.js:1:82279
        at e.tidy (src/ImageClassifier/index.test.js:1:142745)
        at Object.n.value (src/ImageClassifier/index.test.js:1:82256)
        at src/ImageClassifier/index.test.js:1:487368
        at e.tidy (src/ImageClassifier/index.test.js:1:142745)
```

Having dug into this a bit, my best guess is that it's a problem with 2 instances of Tensorflow, so that there's a check in Tensorflow that does `obj instanceof Tensor` and it's comparing it to the wrong copy of tensorflow's Tensor. But I haven't been able to confirm this - it really does look like there's one copy of tfjs-core in the mix.

</details>

### Things this doesn't resolve but I've noticed

- pitchDetection almost certainly has a race condition - its constructor loads a tf model asynchronously but nothing waits for that loading to end, and methods like `processMicrophoneBuffer` could conceivably run before `this.model` is defined / usable.
- poseNet's implementation doesn't appear to precisely match its docs - at least in my reading of the code, it'll call its callback for each frame with a new prediction, rather than calling the callback once with an array.
- I'm pretty sure that the [YOLO documentation](https://ml5js.org/docs/YOLO) is also off by a bit: creating a YOLO object without a video parameter is not valid - i always calls `this.loadVideo()` right now.
- It appears that YOLO only works for 416x416 images?

### Callback situation

Right now this doesn't attempt to change or make the callback system uniform, because the conversation in #143 is still ongoing. I don't anticipate a lot of code change based on the outcome of that discussion, it's mostly a design problem rather than a technical challenge.

tmcw commented:
I'm going to guess that CircleCI is unhappy because of a WebGL support issue? Tests pass locally for me.

cvalenzuela commented:
This is great, thanks @tmcw!

I'm not super familiar with running tests with WebGL support. Maybe @nsthorat might help with this?

nsthorat commented:
I would user browserstack via travis! They are phenomenal and free for small projects (it's still free for us), they actually run your tests on real devices with WebGL enabled.

Here is our travis config which kicks off our tests:
https://github.com/tensorflow/tfjs-core/blob/master/.travis.yml#L6

Karma, the test runner, gets kicked off here:
https://github.com/tensorflow/tfjs-core/blob/master/scripts/test-travis.sh

Here is the karma config, which has info about browserstack as well as local testing:
https://github.com/tensorflow/tfjs-core/blob/master/karma.conf.js

Through travis, you add the environments for your browserstack key, which you can see above.

Let me know if you need any help!

tmcw commented:
Okay, so - moving along with some of this:

- The karma setup could really use some sourcemaps - right now errors in tests all come from the same line of the same file - so I'm trying to add that in this PR, with... unfortunately pretty limited success, karma seems to hang once sourcemaps are added. Any karma whizzes, help would be appreciated; there aren't many clues for debugging, the browser page opens but is blank, no console errors.
- Looks like the answer to that question of promises or callbacks is - with the API as it is currently, it would be necessary to return the object if a callback is provided and a promise if it isn't - because right now the callbacks don't have any argument values.

cvalenzuela commented:
@oveddan helped us with some setup of karma and sourcemaps, he might help us on this. 

As for the promises/callback, I think that's the right approach. Return the object is a callback is provided and Promise if not.

tmcw commented on src/LSTM/index.js:
```diff
@@ -76,83 +74,86 @@ class LSTM {
     const temperature = +options.temperature || this.defaults.temperature;
     const results = [];
 
-    if (this.ready) {
```
I think this is a bugfix - the previous behavior is that calling this method before the model is fully loaded would do nothing.

cvalenzuela replied:
yep, thanks for the catch!

tmcw commented on src/PitchDetection/index.js:
```diff
@@ -16,12 +16,12 @@ class PitchDetection {
     this.audioContext = audioContext;
     this.stream = stream;
-    this.loadModel();
+    this.ready = this.loadModel().then(() => this);
```
I'd also posit that this is a bugfix; pitchdetection has an async loading step but no callback.

cvalenzuela replied:
👍

meiamsome replied:
Note that this doesn't actually add a callback to the constructor.

tmcw commented on src/utils/checkpointLoader.js:
```diff
@@ -10,61 +10,39 @@ export default class CheckpointLoader {
   getVariable(varName) {
     if (!(varName in this.checkpointManifest)) {
-      throw new Error(`Cannot load non-existant variable ${varName}`);
+      throw new Error(`Cannot load non-existent variable ${varName}`);
```
minor typo fix here.

cvalenzuela replied:
👍

tmcw commented on src/utils/checkpointLoader.js:
```diff
@@ -10,61 +10,39 @@ export default class CheckpointLoader {
-    });
+  async loadManifest() {
+    try {
+      this.checkpointManifest = await (await fetch(this.urlPath + MANIFEST_FILE)).json();
```
possibly there was some reason why this bit of code used XMLHttpRequest? It's not apparent in the source before, and I think it'd be ideal if `fetch` is used uniformly.

nsthorat replied:
This was probably taken from the old checkpoint loader in deeplearnjs, there's likely no reason not to use fetch.

cvalenzuela replied:
yes, we took this from the previous checkpoint loader. thanks for the update.

tmcw commented:
Okay - we're getting there! This PR now includes promisification for all of the basic methods in ml5, and I've released a scoped testing module which works pretty well: https://beta.observablehq.com/d/d421a46185d4f063 This also confirms (phew) that Mobilenet isn't actually broken in general or anything - it's just being cranky about the testing environment.

The main tricky one is **PoseNet**:

The docs for PoseNet describe callback as `A function to run once the model has been loaded.` - which is definitely pretty far from what it actually does - the callback is called every time that poses are gotten from a video feed. The actual behavior is kind of tricky: If you don't provide a callback initially, then getting those poses is not obvious - it really looks like you'll end up with two `tf.nextFrame()` loops.

The most common abstraction for this is an event emitter:

```js
let pn = await poseNet(video);
pn.on('pose', pose => {
   // do something with pose
});
```

Because really there are two stages here:

1. The asynchronous setup stage of posenet
2. The many poses that are emitted from it

My take is that we'd want to separate these and make the first one a promise or callback, and make the second one an event. But I'm happy to implement another thing if it fits better into the p5/ml5 ecosystem.

This is ready for a directional review if any folks have the time to spare!

cvalenzuela commented:
This is a really amazing work @tmcw! thanks for all your work and help! 

Since these are a lot of changes and some might affect the API, we will possible merge this after the beta release. I hope to go over all the changes more throughly tomorrow and also get some feedback from @shiffman.

shiffman commented:
Wow, this is fantastic! @cvalenzuela, yes, I think a discussion at a working session would be great. Thank you @tmcw, we are so grateful for your contribution!

oveddan commented:
My two cents: if you want to use promises/async, why not just call the original posenet api?

nsthorat commented on src/FeatureExtractor/Mobilenet.js:
```diff
@@ -115,38 +118,33 @@ class Mobilenet {
+      if (this.usageType === 'classifier') {
+        y = tf.tidy(() => tf.oneHot(tf.tensor1d([label], 'int32'), this.numClasses));
+      } else if (this.usageType === 'regressor') {
+        y = tf.tidy(() => tf.tensor2d([[label]]));
```
no need for tidy() here this actually won't do anything (there are no intermediate tensors)

nsthorat commented on src/ImageClassifier/index.test.js:
```diff
@@ -1,50 +1,44 @@
-  epochs: 20,
-  numClasses: 2,
-  batchSize: 0.4,
+import imageClassifier from './index';
```
cc @cvalenzuela one really useful thing to unit tests is memory usage.

One pattern we do is something like this:

```
const numTensors = tf.memory().numTensors;
apiCall();
expect(tf.memory().numTensors).toEqual(numTensors);
```

This makes sure you don't introduce leaks.

cvalenzuela replied:
great!, I didn't knew that.

nsthorat commented on src/YOLO/index.test.js:
```diff
@@ -0,0 +1,31 @@
+  let yolo;
+
+  beforeEach(async () => {
+    jasmine.DEFAULT_TIMEOUT_INTERVAL = 10000;
```
cc @cvalenzuela you may want to mock the actual fetch call in your unit tests so your tests are hermetic

cvalenzuela replied:
can you expand a bit on this? I'm not sure what hermetic means in the context.

nsthorat commented on src/LSTM/index.js:
```diff
@@ -29,130 +29,131 @@ class LSTM {
-        LSTMCells.push(lstm(i));
-      }
+    await this.ready;
+    const forgetBias = tf.tensor(1.0);
```
@cvalenzuela here and below the tensors are leaking. Same comment above, I'd add a unit test that asserts generate() doesn't create more tensors than you expect.

cvalenzuela replied:
should we wrap this in `tf.tidy()`?

nsthorat commented on src/FeatureExtractor/Mobilenet.js:
```diff
@@ -57,17 +56,23 @@ class Mobilenet {
   }
 
-  classification(video, callback) {
+  classification(video, callback = () => {}) {
```
why not promise-ify these as well?

meiamsome commented:
I'm not sure how I feel about the `callback = () => {}` and then not checking the value of callback later pattern - you might end up in a case where someone passes in a `null` to this function (Perhaps to add an errorCallback where these are supported without a regular callback, or as part of a wrapping function)

tmcw commented:
@meiamsome I haven't converted this to the 'two callbacks pattern' or the 'node-style error argument' pattern because that discussion is ongoing. Regardless, if the two callbacks pattern is adopted, they could use `undefined` and it would properly fall back to the optional value. ml5/p5's APIs are very tolerant, I suppose, so maybe it should _also_ check whether the value is falsy, but that's really a style decision for the maintainers.

cvalenzuela commented:
Sorry for the delay in reviewing this @tmcw. I'm still learning about best practices and patterns with await/ES6. 

My personal take with the callback pattern is to go with the conventional JS style. Since this would play better with more JS libraries. Although I should mention that we have been also discussing a more formal p5 support (https://github.com/ml5js/ml5-library/issues/92). That support might change the callback pattern conversation a bit. I hope this will be something we can resolve in the next few days. 

Regarding the other comments:
 - `pitchDetection` needs some clean up and changes. We are tracking it here https://github.com/ml5js/ml5-library/issues/158 and @handav was working on it. Once this PR is merged we should also update that method to follow the pattern we take here.
- `poseNet` will also get a general clean up and better usage with the original api. 
- YOLO does only work with 416x416 videos for now. That's why that class creates a custom video element.
- After this gets merged, I will pushed another PR with better testing/docs, Browserstack support and a few changes following @nsthorat recommendations.

thanks!

tmcw commented:
Okay!

- Implemented the (error, callback) pattern everywhere I could find, implementing a tiny helper method `callCallback` to cut down on the boilerplate. This PR is about +45loc net, which isn't that bad.
- PoseNet now extends EventEmitter and emits `pose` events.
- Added a small test for LSTMGenerator that will pass once the testing environment is tensorflow-friendly.

This doesn't yet include the bits about counting tensors and memory leaks because, unfortunately, tensorflow is not yet happy in the Karma setup. Besides that, ready for a re-review.

nsthorat commented:
Just curious, what is not happy in the karma setup?

tmcw commented:
How it manifests currently is:

```
    Error: Argument 'x' passed to 'conv2d' must be a Tensor, but got object.
        at c (src/ImageClassifier/index.test.js:1:2730)
        at a (src/ImageClassifier/index.test.js:1:2184)
        at n (src/ImageClassifier/index.test.js:1:2388)
        at Object.i (src/ImageClassifier/index.test.js:1:2413)
        at e.conv2d (src/ImageClassifier/index.test.js:1:24595)
        at src/ImageClassifier/index.test.js:1:81536
        at e.tidy (src/ImageClassifier/index.test.js:1:141252)
        at Object.n.value (src/ImageClassifier/index.test.js:1:81513)
        at src/ImageClassifier/index.test.js:1:480151
        at e.tidy (src/ImageClassifier/index.test.js:1:141252)
```

Or for the LSTMGenerator:

```
      Chrome 67.0.3396 (Mac OS X 10.13.5)
    Error: Argument 'data' passed to 'multiRNNCell' must be a Tensor, but got object.
        at c (src/LSTM/index.test.js:1:2730)
        at a (src/LSTM/index.test.js:1:2184)
        at n (src/LSTM/index.test.js:1:2388)
        at Object.i (src/LSTM/index.test.js:1:2413)
        at e.multiRNNCell (src/LSTM/index.test.js:1:38789)
        at src/LSTM/index.test.js:1:81536
        at e.tidy (src/LSTM/index.test.js:1:141252)
        at Object.n.value (src/LSTM/index.test.js:1:81513)
        at Object.generateInternal (src/LSTM/index.test.js:1:676000)
```

I did a little digging trying to figure out what could be at fault, and my best guess (being an expert in neither tfjs nor karma) is that something in the test build process is creating two copies of tfjs, so tensorflow's instanceof checks are failing and rejecting objects created from a separate copy? Trying out the same code as in the tests but with `yarn start` / on a dist of ml5 works perfectly well, it's only in the test environment that these things fail.

nsthorat commented:
Ah I think your guess sounds right. I looked at the code and don't see anything obviously wrong.

Can you try updating all of the versions of tensorflow / models in package.json and make the 3 TensorFlow deps peerDependencies?

If that doesn't work, I can try to pull the library down after you merge and investigate.

nsthorat commented:
Oh and try loosening the dependencies to use "~" instead of exact versions.

tmcw commented:
Yep, the only clue i'm really finding is:

```
warning "@tensorflow/tfjs > @tensorflow/tfjs-converter@0.4.1" has incorrect peer dependency "@tensorflow/tfjs-core@0.11.3".
```

Besides that, updating the deps doesn't resolve the issue, and looking at yarn.lock and digging in node_modules doesn't turn up any duplicates - there appears to be only one resolved version of tfjs in this branch and in the main branch.

nsthorat commented:
Cool -- I'll fix that ASAP and see if it helps (and write a regression test :) )

cvalenzuela commented on src/LSTM/index.js:
```diff
@@ -29,130 +30,130 @@ class LSTM {
 
-const LSTMGenerator = (modelPath = './', callback = () => {}) => new LSTM(modelPath, callback);
+const LSTMGenerator = (modelPath = './', callback) => {
+  const instance = new LSTM(modelPath);
```
`callback` should also be passed to the instance

cvalenzuela commented on src/LSTM/index.js:
```diff
@@ -29,130 +30,130 @@ class LSTM {
+        generated += mapped;
+      }
+    });
+    return generated;
```
I think this is better than returning and object with `{ generated: '....'  }`

cvalenzuela commented on src/PoseNet/index.js:
```diff
@@ -22,80 +24,74 @@ const DEFAULTS = {
-        tf.nextFrame().then(() => { this.singlePose(callback); });
-      });
+    const pose = await this.net.estimateSinglePose(input, this.imageScaleFactor, this.flipHorizontal, this.outputStride);
+    this.emit('pose', [{ pose, skeleton: this.skeleton(pose.keypoints) }]);
```
What's the benefit of using `EventEmitter`? 

Although, looking at this class again, I think we should maintain consistency across all methods an have the user manage the loop events. So a working example of this should be:

```javascript
// Create the class with no `detectionType` argument
const poseNet = ml5.poseNet(video, modelLoaded);

// Make a prediction using the video
poseNet.singlePose((err, results) => {
  poses = results;
});
```

That way this class also matches the style of the other ml5 methods
```javascript
// Make a prediction using the an image
poseNet.singlePose(img, (err, results) => {
  poses = results;
});
```

tmcw replied:
> What's the benefit of using EventEmitter?

An EventEmitter seems like the right abstraction for PoseNet because:

* It allows the user to both _listen_ (addEventListener/on) and _unlisten_ (removeEventListener/off) from new poses
* It allows multiple listeners to the poses

For example: let's say that you want to show both the poses overlaid on a video element, and those poses as JSON. With an event listener, you'd just call `.on(` with two different listeners. And if want to stop the video, you can use `.off(` to stop listening for poses.

That's why I don't think that this _should_ match the style of other ml5 methods, because it's doing something different. It's totally appropriate to use a callback or a promise if you're calling a method that's expected to have one result. But it's very unusual to use the callback / promise pattern for things that happen based on user input, like frames from a webcam. That's why, for example, we use events for click events on the web.

I think this API is pretty awkward even with the EventEmitter addition: for example, it's not possible to create a PoseNet instance, have it generate single poses, and then switch it to generating multi poses. It's stuck between this idea of immutable models - in which case, the answer would be 'just throw it away and create a new PoseNet', and models that support change - in which case we'd have a single `tf.nextFrame` loop that does an arbitrary 'current setting' on each frame. Currently it _looks_ like you'd be able to call singlePose() and multiPose() and change the PoseNet behavior, but in fact every time you call them you _add_ a new loop and extra attendant computation.

cvalenzuela commented on src/StyleTransfer/index.js:
```diff
@@ -135,15 +131,16 @@ class StyleTransfer extends Video {
 
-  return new StyleTransfer(model, video, callback);
+  const instance = new StyleTransfer(model, video, callback);
+  return cb ? instance : instance.ready;
```
This should be `return callback ? instance : instance.ready;`

cvalenzuela commented on src/YOLO/index.js:
```diff
@@ -161,8 +155,8 @@ const YOLO = (videoOrOptionsOrCallback, optionsOrCallback, cb = () => {}) => {
 
-  return new YOLOBase(video, options, callback);
+  const instance = new YOLOBase(video, options, callback);
+  return cb ? instance : instance.ready;
```
This should be ` return callback ? instance : instance.ready;`

cvalenzuela commented on src/YOLO/index.js:
```diff
@@ -43,108 +43,102 @@ class YOLOBase extends Video {
-    }
-    console.warn('Model has not finished loading');
-    return false;
+    await tf.nextFrame();
```
I think we should move this to the beginning of the function, so to only run over the predicions when the `tf` is ready. This made my browser crash.

cvalenzuela commented:
I'm getting the same errors when running the tests. Loosening the dependencies didn't worked.

I also left some comments. 

My major concern with Promise/Callbacks discussion, is that I think we should support the callback functionality, even when no callback is passed as argument. This is what I mean:

You should still be able to use the `imageClassifier()` like this:
```javascript
const classifier = ml5.imageClassifier('MobileNet');
classifier.predict(img, gotResult);
```
The same will be achieve if a callback is passed to the constructor
```javascript
const classifier = ml5.imageClassifier('MobileNet', callback);
classifier.predict(img, gotResult);
```

But if `then()` is used, you can use promises
```javascript
ml5.imageClassifier('MobileNet').then(c => { /* ... */ });
```
Right now, if you don't provide a callback, this fails since it's being constructed as a Promise
```javascript
const classifier = ml5.imageClassifier('MobileNet');
classifier.predict(img, gotResult); // Uncaught TypeError: Cannot read property 'predict' of undefined
```

I'm not sure what's the best approach for something like that or if it is even possible for that pattern to exist. Any thoughts?

tmcw commented:
> I'm not sure what's the best approach for something like that or if it is even possible for that pattern to exist. Any thoughts?

So... well, I don't think that'd be a very consistent pattern in the general sense, mostly because of word2vec. word2vec needs an async step to initially set itself up, and then has sync methods. So, if you don't use a callback, it'll be broken by default - all the methods rely on it already being initialized.

I think that there are a few approaches here:

1. **No async constructors.**  All class methods `await this.ready` before returning a result.
  * Advantage: simpler - less async handling required.
  * Disadvantage: any errors involved with the initial setup step are only returned / passed to a callback once you start using the object.
2. **Make objects thenable.** [Per promises/a+](https://promisesaplus.com/), if we add a `.then()` method to the classes, we can treat them as promises _and_ have them expose methods. Seems fine, though, per the first point... this'll still leave things like word2vec broken by default if someone doesn't use a callback.
3. **Make objects thenable and change the word2vec api.** Combining 1 and 2, we'd make objects thenable _and_ eliminate any remaining sync class methods, to resolve the word2vec issue.

cvalenzuela commented:
**Option 3** seems the best one for me. Although, how much more work will be implementing Promises/A+ in all classes? I'm not familiar with that syntax. If it's too much work I think we should move with **Option 1**

meiamsome commented on src/FeatureExtractor/Mobilenet.js:
```diff
@@ -40,12 +40,8 @@ class Mobilenet {
-      callback();
-    });
+    this.ready = callCallback(this.loadModel(), callback);
+    this.then = this.ready.then;
```
I don't know if this is to spec, as I don't think the `this` is bound by default on a promise. Perhaps `this.ten = this.ready.then.bind(this.ready)`

I am also not sure about using `ready` as the name of the Promise, it seems to me that ready implies a boolean as it is in other places in the library. Perhaps `readyPromise` or `loadingPromise`? I'm not sure but `await model.ready` doesn't read very well to me...

tmcw replied:
I'm fine with any name for this, but note that this PR changes all `ready` properties to be Promises, so there is no boolean implication. I don't think a `Promise` suffix is appropriate; that isn't the pattern anywhere else in JavaScript, like the Response object of a fetch call does not have `.textPromise`, it has `.text()`.

Also, `.ready` is an completely internal property: it isn't super worthwhile to bikeshed the name.

meiamsome replied:
Is it internal though? There's at least one [example using it](https://github.com/ml5js/ml5-examples/blob/9abe6b284c81f80b05f44f828a3a5b551a62ecff/p5js/StyleTransfer_Image/sketch.js#L36). I'm just not a fan of changing exposed APIs that aren't explicitly private. I would possibly be fine with just `_ready` or `readyInternal` such that the private nature of the property is clear. This is something that should be decided for the whole library so that internal properties don't appear to be public. I would personally push for prefixing private properties with `_`, but that's because I use the convention from Python.

tmcw replied:
That example uses the `master` branch. I defer to the maintainers on the rest of this PR.

cvalenzuela commented:
Thanks @tmcw! Although, with the changes @meiamsome introduced in https://github.com/ml5js/ml5-library/pull/173 the test suite should be working. We have been developing the library against examples but we should start relying more on tests now.

I'll be happy to take over this PR if you preferred so and make the necessary changes to get this merged.

tmcw commented:
The word2vec tests were (sometimes) working in this branch too - the hard test failures, the ones that consistently fail, are in ImageClassifier, LSTMGenerator, Mobilenet, and the other modules.

cvalenzuela commented:
I think they were not working because there were multiple `tf` versions in karma. @meiamsome tweaked the karma config and figure that doing:
```javascript
const { tf, word2vec } = ml5;
```
instead of 
```javascript
import word2Vec from './index';
```
fixed the issue.

meiamsome commented on src/utils/callcallback.js:
```diff
@@ -0,0 +1,14 @@
+// This software is released under the MIT License.
+// https://opensource.org/licenses/MIT
+
+export default function callCallback(promise, callback) {
```
I'm not sold on `callCallback` - that requires handling a promise internally within the `async` functions - something like `return callCallback(Promise.resolve(result), callback);` could theoretically be used to return a result that is non-promised but I think you lose a lot of the wonder of an `async` function by doing this. As I suggested in #143 a wrapping function that creates callbacks from the `async` Promises would probably be my preferred technique - I have a working suggestion that's updated now here for decorators: https://github.com/ml5js/ml5-library/compare/master...meiamsome:testing/promise-to-callback-decorator or here for with class properties: https://github.com/ml5js/ml5-library/compare/master...meiamsome:testing/promise-to-callback-class-properties, although both of them need to used babel-eslint because the features are not fully standardised they both give the benefits that the functionality of the Promise and the callback(s) are intrinsically linked. Obviously, if supporting one callback type is required it would significantly reduce the complexity of the wrapping fn/decorator.

tmcw replied:
Is there any specific behavior in the exposed API surface, in this branch, that seems incorrect? I'm finding it hard to parse from your comment whether the issue is code aesthetics or if there is a specific method behavior that would be different if we were to go with a different approach.

meiamsome replied:
I have gone over the PR again and am fairly sure that the exposed* API is mostly correct - my main concern is an aesthetic one here, but there are some other important things such as the fact any errors inside the functions are just never sent to the callback function at the moment, and adding it in is a matter of wrapping almost every function in a try/catch statement to send it correctly at that least, which is a lot of boilerplate when that could be handled by wrapping the function itself. Also, the if/else if/.../else chaining for parsing arguments that's just a big mess. 

* In addition, there's some `async` functions that don't allow callbacks but I'm not sure these are public or private (see the other discussion on private properties for that, I guess)

tmcw commented:
All right - 'allow edits from maintainers' is on - off to you, maintainers!

cvalenzuela commented:
thanks @tmcw! we will take it from here!
